src/runtime/metadata.cpp

```diff
@@ -1,10 +1,56 @@
 #include "HalideRuntime.h"
+#include "scoped_mutex_lock.h"
 
-// This is a trick used to ensure that halide_filter_metadata_t (and related types)
-// are included in the runtime bitcode; since no (other) runtime function references them
-// they are ordinarily stripped, making Codegen_LLVM fail when we attempt to access
-// the types.
+extern "C" {
 
-#define INLINE inline __attribute__((weak)) __attribute__((used)) __attribute__((always_inline)) __attribute__((nothrow)) __attribute__((pure))
+struct _halide_runtime_internal_registered_filter_t {
+    struct _halide_runtime_internal_registered_filter_t *next;
+    const char *name;
+    const halide_filter_metadata_t* metadata;
+    int (*argv_func)(void **args);
+};
 
-INLINE void __force_include_halide_filter_metadata_t_types(halide_filter_metadata_t*) {}
+};
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+struct list_head_t {
+    halide_mutex mutex;
+    _halide_runtime_internal_registered_filter_t *next;
+
+    list_head_t() : mutex(), next(NULL) {}
+};
+
+static list_head_t* get_list_head() {
+    static list_head_t head;
+    return &head;
+}
+
+} } }
+
+extern "C" {
+
+// This is looked up by name in Codegen_LLVM, which is easier to do
+// for functions with plain C linkage.
+WEAK void halide_runtime_internal_register_metadata(_halide_runtime_internal_registered_filter_t *info) {
+    // Note that although the metadata pointer itself is valid, the contents pointed
+    // to by it may not be initialized yet (since order of execution is not guaranteed in this case);
+    // it is essential that this code not do anything with that pointer other than store
+    // it for future use. (The name argument will always be valid here, however.)
+    list_head_t* head = get_list_head();
+    ScopedMutexLock lock(&head->mutex);
+    info->next = head->next;
+    head->next = info;
+}
+
+WEAK int halide_enumerate_registered_filters(void *user_context, void* enumerate_context, enumerate_func_t func) {
+    list_head_t* head = get_list_head();
+    ScopedMutexLock lock(&head->mutex);
```

Comment:
Well, I need to test it on Windows anyway...

On Tue, Apr 7, 2015 at 11:08 AM, Steven Johnson <notifications@github.com>
wrote:

> In src/runtime/metadata.cpp
> <https://github.com/halide/Halide/pull/733#discussion_r27904129>:
>
> > +// This is looked up by name in Codegen_LLVM, which is easier to do
> > +// for functions with plain C linkage.
> > +WEAK void halide_runtime_internal_register_metadata(_halide_runtime_internal_registered_filter_t *info) {
> > +    // Note that although the metadata pointer itself is valid, the contents pointed
> > +    // to by it may not be initialized yet (since order of execution is not guaranteed in this case);
> > +    // it is essential that this code not do anything with that pointer other than store
> > +    // it for future use. (The name argument will always be valid here, however.)
> > +    list_head_t* head = get_list_head();
> > +    ScopedMutexLock lock(&head->mutex);
> > +    info->next = head->next;
> > +    head->next = info;
> > +}
> > +
> > +WEAK int halide_enumerate_registered_filters(void *user_context, void* enumerate_context, enumerate_func_t func) {
> > +    list_head_t* head = get_list_head();
> > +    ScopedMutexLock lock(&head->mutex);
>
> Oh. Indirectly. Yes, you're right. You have a good point, I don't know if
> we do this elsewhere.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/733/files#r27904129>.
>